[Detection Engine] ML Rule Alert Suppression - Followup

rylnd opened the pull request:
## Summary

This PR is a followup to #181926. A demo instance can be found [here](https://rylnd-pr-188267-rylnd-ml-suppression-followup.kbndev.co/). It includes the following changes:

- Refactoring some Rule Form logic with `useMemo` 
  - Requested [in this discussion](https://github.com/elastic/kibana/pull/181926#discussion_r1656825268)
  - Addressed in a5fcf4d0ccd738a40746d099bd295484781825f5
- Adds FTR tests validating ML Suppression supports alert enrichment
  - Requested [during previous review](https://github.com/elastic/kibana/pull/181926#discussion_r1634616090)
  - Addressed in d5aa55159070bd716892360f0321505fb44943c0
- Disables ML Suppression fields as a group
  - Requested in [this comment](https://github.com/elastic/kibana/pull/181926#issuecomment-2203592643) 
  - Addressed by 983945b8da1db8f1309db559ec74e5df7c064862
- Only disables ML Suppression UI when no fields info is available
  - After discussing [this behavior](https://github.com/elastic/kibana/pull/188267#discussion_r1684077526) with @vitaliidm , I was going to add a special case for a single job being selected that has field info (and not disabling then), but since those fields are tied specifically to the selected jobs, I see no reason not to base the enablement of that UI purely on the presence/absence of field information (rather than on jobs status, as it was previously). The warning when not all jobs are started will still be displayed. cc @approksiu .
  - Addressed in https://github.com/elastic/kibana/pull/188267/commits/f0bd4b9255d310a9a500ec38757ccf5a5e84920f.


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))



<!--ONMERGE {"backportTargets":["8.15"]} ONMERGE-->

rylnd committed:
Refactor nested ternary logic into useMemo fns

It would be painful to try to write a pure function that captures this
same logic (see the deps lists), but as a compromise we can at least
encapsulate these in functions and use more expressive implementation.

rylnd committed:
Add tests validating enrichment behavior for ML suppression

This was requested during review of #181926, and I'm circling back to
that now.

rylnd committed:
Disable all suppression fields as a group

We did not previously extend our full disabling logic to all suppression
fields. Now, when we determine that alert suppression is
invalid/disabled, we disable all those relevant fields together.

This commit accomplishes the above, and also factors out some shared
boolean logic into the more general `areSuppressionFieldsDisabled`
boolean, which is shared by most of these fields.

rylnd commented:
/ci

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -477,77 +477,90 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
    * -  Eql sequence is used and suppression fields are in the default state
+   * - ML Field information is not available
    * */
   const isDurationDisabled =
```
Duration for threshold rule is not disabled, when checkbox to enabled suppression is not checked

<img width="940" alt="Screenshot 2024-07-17 at 12 06 55" src="https://github.com/user-attachments/assets/5675fb1a-eff1-4978-8c4a-75d57563c3c3">

rylnd replied:
In the default state (just selecting a threshold rule) this doesn't occur, but I was able to reproduce by:

1. Selecting a "group by" field in the Query Rule tab
2. Selecting the Threshold rule type

@vitaliidm I believe this is related to a previous conversation we had: on [this line](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx#L539), `groupByFields.length` won't be nonzero in the above case, because `groupByFields` is not used (nor reset) by threshold rules.

This PR doesn't introduce the behavior we're describing, but we can certainly address it now.

vitaliidm replied:
@rylnd, thanks, would be great if you can fix it in this PR.
Otherwise, I can create a separate issue and address it later

rylnd replied:
@vitaliidm this was addressed in d2282c914a55352f31dfbb44af3ea33d718c703e

vitaliidm replied:
thanks for fixing this one

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -477,77 +477,90 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
+  /**
+   * If we don't have ML field information, users can't meaningfully interact with these fields */
+  const areSuppressionFieldsDisabledByMlFields =
+    isMlRule(ruleType) && (noMlJobsStarted || mlFieldsLoading || !mlSuppressionFields.length);
```
<img width="952" alt="Screenshot 2024-07-17 at 12 05 43" src="https://github.com/user-attachments/assets/7c2dfc98-8325-4a92-8448-ce012ea04a1e">

I added two jobs (one running, one not). Selected field, configured suppression and then removed running job.
So, suppression config is disabled, but field actually was from a running job.

If I save rule, with "disabled" suppression and start job - suppression would work, but the field stored there would be from another job.

 I can't clear this suppression on form , because when I removed job, it is still disabled
<img width="954" alt="Screenshot 2024-07-17 at 12 13 15" src="https://github.com/user-attachments/assets/efa91bb5-3294-4405-9e41-9b99f19100fe">

So, I have to add the same running job to clear it.

Similar can happen when user changes index, without adjusting suppression fields. But in this case, suppression controls are not disabled and user can easily change those fields
Here, after removing running job, controls become disabled. It requires user to add this job again to change fields or start another.
I am not sure, how critical it can be - just an observation on behaviour.
What do you think about it?

rylnd replied:
Thanks for calling this out. I agree that the UX wasn't great, so I fixed it in 8b6c460e3815258a6c8c314ae5822e4d4e4ad485.

rylnd committed:
Fix case where suppression duration input on threshold rules was incorrectly enabled

When a non-threshold suppression field had been selected, the previous
logic was causing the duration input to be enabled, because it was based
on the non-threshold suppression input.

rylnd committed:
Refactor suppression field logic

This commit abstracts a `areSuppressionFieldsSelected` boolean to
capture the case where we can't disable fields due to the user needing
to be able to change them. We then refactored the relevant predicates in
terms of that new boolean.

rylnd committed:
Refactor predicate comments

Now that these are refactored in terms of more general predicates, it
doesn't make much sense to enumerate all these situations for each
variable, nor are these lists particularly helpful to understanding the
code.

So that we only have one source of truth to maintain moving forward (the
code itself), I'm removing these "leaf" comments and instead annotating
their parents.

rylnd commented:
@vitaliidm thanks for the feedback, your comments should now be addressed. Please take another look when you can!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222223)
* Commit: b96ad5d40c2ce2cdfa7d73fe9d179766b7ebc82c

### Failed CI Steps
* [Jest Tests #11](https://buildkite.com/elastic/kibana-pull-request/builds/222223#0190c2b1-7748-49ee-98f4-fbed1c851605)
* [Jest Integration Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/222223#0190c2b1-774d-4a2e-a8b3-dce3be77d8da)
* [Serverless Detection Engine - Security Solution Cypress Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/222223#0190c29c-a697-40ca-b3ff-125dcad3cfcf)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222223#0190c2b1-7748-49ee-98f4-fbed1c851605) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222223/jobs/0190c2b1-7748-49ee-98f4-fbed1c851605/artifacts/0190c2d0-f9e9-4a84-abdb-a6aa74aa2526) Jest Tests #<span></span>11 / Template fields renders template fields with existing value
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222223#0190c29c-a697-40ca-b3ff-125dcad3cfcf) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222223/jobs/0190c29c-a697-40ca-b3ff-125dcad3cfcf/artifacts/0190c2d4-f13c-4798-a890-f7175a7af358) Serverless Detection Engine - Security Solution Cypress Tests #<span></span>4 / Threshold rules Creates a new threshold rule with suppression enabled Creates a new threshold rule with suppression enabled

cc @rylnd

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222223","number":222223,"commit":"b96ad5d40c2ce2cdfa7d73fe9d179766b7ebc82c"}],"number":222223}
buildkite-pr-comment-->

rylnd commented:
@vitaliidm hold off on that review. As the latest build shows, I've still got some bugs 👀

rylnd committed:
Fix errors in suppression field enablement logic

Besides `areSuppressionFieldsSelected` not being defined as the comment
documents, we had an issue where threshold suppression fields were not
properly disabled in the default state (which also had some consequences
for switching between rule types). This is now fixed, and IMO these
predicates now read much better as well.

rylnd committed:
Merge branch 'main' into rylnd/ml_suppression_followup

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -472,82 +472,69 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
-    !groupByFields.length;
+    areSuppressionFieldsDisabled || isThresholdSuppressionDisabled || !areSuppressionFieldsSelected;
+  const isMissingFieldsDisabled = areSuppressionFieldsDisabled || !areSuppressionFieldsSelected;
 
```
**Spike in logon events** is not installed, when selected with running job, no warning is displayed

vitaliidm replied:
@rylnd 

1. Job was created through wizard

https://github.com/user-attachments/assets/9332edfb-d9ed-4177-a74a-f3a1350fc585


2. This is behaviour in main
https://github.com/user-attachments/assets/9e731d43-92d5-48b6-89b6-6ddcbe528047

rylnd replied:
@vitaliidm this is certainly an interesting one that I hadn't encountered. I'm still not able to reproduce; after creating a custom job (and getting it in a persistent "started" state), I see the expected behavior:

https://github.com/user-attachments/assets/72fa012c-d4a6-4ca9-a83a-586bfac6b4dc

And you can see in the following screenshot that the custom job shows up with the rest of the jobs and is otherwise treated no differently (which was my previous theory on this behavior): 

<img width="1840" alt="Screenshot 2024-07-19 at 4 48 39 PM" src="https://github.com/user-attachments/assets/feb6e21a-b2ae-4325-ae28-4e9023dbd98b">

So perhaps there's something in your job's state that's different than what I'm seeing, or our environments are otherwise different? Let's pair on Monday and figure it out.

rylnd committed:
Only disable ML Suppression UI if no fields are available

We continue to show a warning if not all of the selected jobs are
running, but we no longer disable the UI based purley on whether none of
the selected jobs are running; it's now based purely on whether we have
fields to display.

rylnd committed:
Rename hook variable to simplify form logic

I had previously used the concept of "some jobs started" because that is
the case that we want to warn users about, and it was conceptually
simpler than "not all jobs started."

However, from both the perspective of the hook and the consumer, it
makes more sense to return an "all jobs started" boolean and use that as
necessary.

This is logically equivalent to the previous commit, just nicer to read.

rylnd committed:
Merge branch 'main' into rylnd/ml_suppression_followup

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/188907">](https://github.com/elastic/kibana/pull/188907)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

vitaliidm commented:
Issue has been discovered: ML warning appears at other rule types 
<img width="1190" alt="Screenshot 2024-07-23 at 10 33 35" src="https://github.com/user-attachments/assets/2b14ed26-4f3a-48a7-a2a5-b7870e8950a9">
